libsql-server/src/http/user/db_factory.rs

```diff
@@ -46,18 +47,25 @@ pub fn namespace_from_headers(
         return Ok(NamespaceName::default());
     }
 
-    let host = headers
-        .get("host")
-        .ok_or_else(|| Error::InvalidHost("missing host header".into()))?
-        .as_bytes();
-    let host_str = std::str::from_utf8(host)
-        .map_err(|_| Error::InvalidHost("host header is not valid UTF-8".into()))?;
+    headers
+        .get(NAMESPACE_METADATA_KEY)
+        .ok_or(Error::InvalidNamespace)
+        .and_then(|h| h.to_str().map_err(|_| Error::InvalidNamespace))
```

Comment:
Wouldn't it be easier/cleaner to do:
```
headers
    .get(NAMESPACE_METADATA_KEY)
    .and_then(|h| h.to_str())
    .ok_or(Error::InvalidNamespace)
```
?

Reply:
I don't understand this whole block of code though. It contains errors but at the end it has `.or_else` which ignores those errors:

```
headers
        .get(NAMESPACE_METADATA_KEY)
        .ok_or(Error::InvalidNamespace) // <=== 1. translating any error to InvalidNamespace
        .and_then(|h| h.to_str().map_err(|_| Error::InvalidNamespace)) // <=== 2. possibly InvalidNamespace if header value can't be translated to string
        .and_then(|n| NamespaceName::from_string(n.into()))
        .or_else(|_| { // <=== ignoring both error from (1) and from (2)
            let host = headers
                .get("host")
                .ok_or_else(|| Error::InvalidHost("missing host header".into()))?
                .as_bytes();
            let host_str = std::str::from_utf8(host)
                .map_err(|_| Error::InvalidHost("host header is not valid UTF-8".into()))?;

            match split_namespace(host_str) {
                Ok(ns) => Ok(ns),
                Err(_) if !disable_default_namespace => Ok(NamespaceName::default()),
                Err(e) => Err(e),
            }
        })
```
What do I miss? How are `.ok_or(Error::InvalidNamespace)` and `.map_err(|_| Error::InvalidNamespace)` changing anything if the error gets ignore at the end by `.or_else(|_| {` step?

Reply:
yea, this code needs to change
need to define explicit rule of when to use namespace from header and when from hostname
@LucioFranco @haaawk 
what do you think about this?
* If get(NAMESPACE_METADATA_KEY) gives Some, use the header.
* Do not fallback to namespace from hostname in case of malformed header, just return Error
* If get(NAMESPACE_METADATA_KEY) gives None, use namespace from hostname
* If disable_default_namespace not set and any of the above fails, return default namespace
* if disable_default_namespace is set, any error produced by the above should be returned to caller

Reply:
committed a new proposal

Reply:
The proposal sounds good to me. You said that you committed it but I don't see code in `namespace_from_headers` being changed, is it?

Reply:
pushed